File: demo/java/custom-java/src/main/java/com/baidu/bce/driver/Simulator.java
package com.baidu.bce.driver;

import com.baidu.bce.models.Point;
import com.baidu.bce.utils.L;

import java.util.Random;
import java.util.concurrent.TimeUnit;

public class Simulator {
    /**
     * SIMULATOR_INTERVAL: 10 seconds
     */
    private static final int SIMULATOR_INTERVAL = 10;
    private String name;
    private Point point;
    private Random random;

    public Simulator(String name) {
        this.random = new Random(System.currentTimeMillis());

        this.name = name;
        this.point = new Point();
        this.point.setTemperature(this.random.nextInt(100));
        this.point.setHumidity(this.random.nextInt(100));
        this.point.setPressure(this.random.nextInt(100));

        generateSimulateData();
    }

    public void set(int index, float val) {
        switch (index) {
            case Point.INDEX_TEMPERATURE -> this.point.setTemperature(val);
            case Point.INDEX_HUMIDITY -> this.point.setHumidity(val);
            case Point.INDEX_PRESSURE -> this.point.setPressure(val);
        }
    }

    public Object get(int index) throws Exception {
        switch (index) {
            case Point.INDEX_TEMPERATURE -> {
                return this.point.getTemperature();
            }
            case Point.INDEX_HUMIDITY -> {
                return this.point.getHumidity();
            }
            case Point.INDEX_PRESSURE -> {
                return this.point.getPressure();
            }
        }
        throw new Exception("unsupported point information");
    }

    private void generateSimulateData() {
        new Thread(() -> {
            while (true) {
                this.point.setTemperature(this.random.nextInt(100));
                this.point.setHumidity(this.random.nextInt(100));
                this.point.setPressure(this.random.nextInt(100));

                L.debug("Simulate generate random temperature = " + this.point.getTemperature());
                L.debug("Simulate generate random humidity = " + this.point.getHumidity());
                L.debug("Simulate generate random pressure = " + this.point.getPressure());
                try {
                    TimeUnit.SECONDS.sleep(SIMULATOR_INTERVAL);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }).start();
    }
}

File: demo/cxx/custom-cxx/src/driver/simulator.cpp
#include "simulator.h"
#include <cstdlib>

Simulator::Simulator(const std::string &n)
        : name(n) {
    srand((unsigned int) time(nullptr));
    point.setTemperature(static_cast<float>(rand() % 100));
    point.setHumidity(static_cast<float>(rand() % 100));
    point.setPressure(static_cast<float>(rand() % 100));

    generateSimulateData();
}

void Simulator::set(int index, float val) {
    switch (index) {
        case Point::INDEX_TEMPERATURE:
            point.setTemperature(val);
            break;
        case Point::INDEX_HUMIDITY:
            point.setHumidity(val);
            break;
        case Point::INDEX_PRESSURE:
            point.setPressure(val);
            break;
        default:
            return;
    }
}

float Simulator::get(int index) {
    switch (index) {
        case Point::INDEX_TEMPERATURE:
            return point.getTemperature();
        case Point::INDEX_HUMIDITY:
            return point.getHumidity();
        case Point::INDEX_PRESSURE:
            return point.getPressure();
        default:
            return 0.0f;
    }
}

void Simulator::generateSimulateData() {
    std::thread([this]() {
        while (true) {
            set(Point::INDEX_TEMPERATURE, static_cast<float>(random() % 100));
            set(Point::INDEX_HUMIDITY, static_cast<float>(random() % 100));
            set(Point::INDEX_PRESSURE, static_cast<float>(random() % 100));

            L::debug("Simulate generate random temperature = " + std::to_string(get(Point::INDEX_TEMPERATURE)));
            L::debug("Simulate generate random humidity = " + std::to_string(get(Point::INDEX_HUMIDITY)));
            L::debug("Simulate generate random pressure = " + std::to_string(get(Point::INDEX_PRESSURE)));

            std::this_thread::sleep_for(std::chrono::seconds(SIMULATOR_INTERVAL));
        }
    }).detach();
}
